### ISSUE TITLE
During heavy write the checkpoint may hang forever on cond wait

### ISSUE THREAD

**Issue by absolute8511** [cockroachdb/pebble#743]
Hi, recently I am doing some tests on the pebble to compare with the rocksdb, however, I encountered a checkpoint issue that will hang forever and it can be reproduced by running a benchmark.

I open the pebble storage as below:
```
func GetRocksdbUint64(v []byte) (uint64, error) {
        if v == nil || len(v) == 0 {
		return 0, nil
	} else if len(v) != 8 {
		return 0, errIntNumber
	}
	return binary.LittleEndian.Uint64(v), nil
}

type Uint64AddMerger struct {
	buf []byte
}

func (m *Uint64AddMerger) MergeNewer(value []byte) error {
	cur, err := GetRocksdbUint64(m.buf, nil)
	if err != nil {
		return err
	}
	vint, err := GetRocksdbUint64(value, nil)
	if err != nil {
		return err
	}
	nv := cur + vint
	if m.buf == nil {
		m.buf = make([]byte, 8)
	}
	binary.LittleEndian.PutUint64(m.buf, nv)
	return nil
}

func (m *Uint64AddMerger) MergeOlder(value []byte) error {
	return m.MergeNewer(value)
}

func (m *Uint64AddMerger) Finish() ([]byte, io.Closer, error) {
	return m.buf, nil, nil
}

func newUint64AddMerger() *pebble.Merger {
	return &pebble.Merger{
		Merge: func(key, value []byte) (pebble.ValueMerger, error) {
			res := &Uint64AddMerger{}
			res.MergeNewer(value)
			return res, nil
		},
		// the name should match the rocksdb default merge name
		Name: "UInt64AddOperator",
	}
}

func OpenPebbleEng(dataDir string) (*pebble.DB, error) {
	err := os.MkdirAll(dataDir, common.DIR_PERM)
	if err != nil {
		return nil, err
	}
	lopts := make([]pebble.LevelOptions, 0)
	for l := 0; l < numOfLevels; l++ {
		compress := pebble.SnappyCompression
		if l <= 1 {
			compress = pebble.NoCompression
		}
		filter := bloom.FilterPolicy(10)
		opt := pebble.LevelOptions{
			Compression:    compress,
			BlockSize:      1024*8,
			TargetFileSize: 64*1024*1024,
			FilterPolicy:   filter,
		}
		opt.EnsureDefaults()
		lopts = append(lopts, opt)
	}

	opts := &pebble.Options{
		Levels:                      lopts,
		MaxManifestFileSize:         32*1024*1024,
		MemTableSize:               64*1024*1024,
		MemTableStopWritesThreshold: 8,
		LBaseMaxBytes:               256*1024*1024,
		L0CompactionThreshold:       8,
		MaxOpenFiles:                -1,
		MaxConcurrentCompactions:    8,
	}

	opts.Merger = newUint64AddMerger()
	cache := pebble.NewCache(1024*1024*512)
	defer cache.Unref()
	opts.Cache = cache
	eng, err := pebble.Open(dataDir, opts)
	if err != nil {
		return nil, err
	}

	return eng, nil
}

```

And in a loop goroutine, I took the checkpoint period.
```
for  {
// do some wait
eng.Checkpoint(path)
....
// other things

}
```

And do heavy write in another goroutine.
```
wb := db.NewBatch()
wb.Set(key, value, writeOpts)
wb.Merge(key, value, writeOpts)
eng.Apply(wb, writeOpts)
```

After running for a while, the Checkpoint hang on below : 
```
goroutine 939 [sync.Cond.Wait, 31 minutes]:
runtime.goparkunlock(...)
        /data/users/xxx/go/src/runtime/proc.go:310
sync.runtime_notifyListWait(0xc00014d588, 0xc000000002)
        /data/users/xxx/go/src/runtime/sema.go:510 +0xf8
sync.(*Cond).Wait(0xc00014d578)
        /data/users/xxx/go/src/sync/cond.go:56 +0x9d
github.com/cockroachdb/pebble.(*DB).disableFileDeletions(0xc00014cd00)
        /data/gopath/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200610205245-cb9c3090f417/compaction.go:1825 +0x3b
github.com/cockroachdb/pebble.(*DB).Checkpoint(0xc00014cd00, 0xc468926b40, 0x50, 0x0, 0x0)
        /data/gopath/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200610205245-cb9c3090f417/checkpoint.go:34 +0x103
        /data/gopath/src/github.com/youzan/ZanRedisDB/engine/pebble_eng.go:451 +0x101
github.com/youzan/ZanRedisDB/rockredis.(*RockDB).backupLoop.func1(0xc4f3c7e640, 0xc004bf8000)
```

It seems some cond signal missed to wake up the cond.Wait.  I am not sure is this a bug or some wrong usage? If more details need, I can provide more information to reproduce.
Thanks.

**Comment by petermattis** [comment]
@absolute8511 I don't see anything obviously wrong with the code in question. Can you reproduce this problem? If yes, can you provide the full source for a reproduction (ideally in the form of a test). That will make it easier to track down what is going on. If you're able to reproduce but unable to share a reproduction, it would be helpful to see a full goroutine dump rather than only the Checkpoint goroutine.

**Comment by absolute8511** [comment]
a full goroutine dump as below:

[go-stack.txt](https://github.com/cockroachdb/pebble/files/4784121/go-stack.txt)

**Comment by petermattis** [comment]
@absolute8511 Thanks. Unfortunately that didn't shed any more light on what is going on. Can you add some instrumentation to Pebble to help track this down? The `disableFileDeletions` call is spinning waiting for `DB.mu.cleaner.cleaning` to become false. That field is set to `true` in exactly one location and set to `false` in exactly one location. I would try logging a stack trace at both of those locations in order to figure out which `cleaner.cleaning = true` assignment does not have a corresponding `cleaner.cleaning = false` assignment.

**Comment by petermattis** [comment]
Actually, I think I see the problem. Can you try running with this diff:

```diff
diff --git a/compaction.go b/compaction.go
index b6c476e3..a0dfb3bc 100644
--- a/compaction.go
+++ b/compaction.go
@@ -1824,7 +1824,7 @@ func (d *DB) disableFileDeletions() {
        for d.mu.cleaner.cleaning {
                d.mu.cleaner.cond.Wait()
        }
-       d.mu.cleaner.cond.Signal()
+       d.mu.cleaner.cond.Broadcast()
 }

 // enableFileDeletions enables previously disabled file deletions. Note that if
@@ -1866,7 +1866,7 @@ func (d *DB) acquireCleaningTurn(waitForOngoing bool) bool {
 // d.mu must be held when calling this.
 func (d *DB) releaseCleaningTurn() {
        d.mu.cleaner.cleaning = false
-       d.mu.cleaner.cond.Signal()
+       d.mu.cleaner.cond.Broadcast()
 }

 // deleteObsoleteFiles deletes those files that are no longer needed.
```

If there are multiple goroutines waiting on a cond, it isn't safe to use `Cond.Signal()` which will only wake one of them. Instead we need to use `Cond.Broadcast`.
